File: FureyLib/C#/Templates/State Machine/StateMachine.cs

// Template State Machine Script
// by Kyle Furey

/// <summary>
/// State machine states enum
/// </summary>
public enum StateType
{
    None = 0,
    State = 1
};

/// <summary>
/// Base for building an state machine.
/// </summary>
public class StateMachine
{
    /// <summary>
    /// The current state of this state machine
    /// </summary>
    public StateBase currentState = null;

    /// <summary>
    /// The current state of this state machine
    /// </summary>
    public States currentStateType = StateType.None;

    /// <summary>
    /// Starting state constructor
    /// </summary>
    /// <param name="newState"></param>
    public StateMachine(StateBase newState)
    {
        SwitchState(newState);
    }

    /// <summary>
    /// Properly switches the state machine's current state
    /// </summary>
    /// <param name="newState"></param>
    public void SwitchState(StateBase newState)
    {
        // Exit the current state
        if (currentState != null)
        {
            currentState.OnStateExit();
        }

        // Switch the current state to the new state
        currentState = newState;

        // Enter the new state
        if (currentState != null)
        {
            currentState.OnStateEnter();
        }
    }

    /// <summary>
    /// Updates the current state (should be called each tick)
    /// </summary>
    public void Tick()
    {
        // Call the current state's update function
        if (currentState != null)
        {
            currentState.OnStateUpdate();
        }
    }
}

/// <summary>
/// The base class for each state. All states should inherit from this class.
/// </summary>
public abstract class StateBase
{
    /// <summary>
    /// The inherited state machine from the owner
    /// </summary>
    protected StateMachine stateMachine = null;

    /// <summary>
    /// Called when this state is set as the state machine's current state
    /// </summary>
    public virtual void OnStateEnter()
    {
        // Note: Logic applies to all inherited states
    }

    /// <summary>
    /// Called every frame while this state is the state machine's current state
    /// </summary>
    public virtual void OnStateUpdate()
    {
        // Note: Logic applies to all inherited states
    }

    /// <summary>
    /// Called when this state machine's current state is no longer this state
    /// </summary>
    public virtual void OnStateExit()
    {
        // Note: Logic applies to all inherited states
    }
}

File: FureyLib/Python/Templates/State Machine/state_machine.py
# Template State Machine Script
# by Kyle Furey

# Include this heading to use the class
# from state_machine import StateBase, StateMachine


class StateBase:
    """The base class for each state. All states should inherit from this class."""

    state_machine: 'StateMachine' = None
    """The inherited state machine from the owner"""

    def on_state_enter(self):
        """Called when this state is set as the state machine's current state"""

        # Note: Logic applies to all inherited states

        pass

    def on_state_update(self):
        """Called every frame while this state is the state machine's current state"""

        # Note: Logic applies to all inherited states

        pass

    def on_state_exit(self):
        """Called when this state machine's current state is no longer this state"""

        # Note: Logic applies to all inherited states

        pass

    pass


class StateMachine:
    """Base for building a state machine."""

    current_state: StateBase = None
    """The current state of this state machine"""

    def __init__(self, new_state: StateBase):
        """Starting state constructor"""

        self.switch_state(new_state)

        pass

    def tick(self):
        """Updates the current state (should be called each tick)"""

        # Call the current state's update function
        if self.current_state is not None:
            self.current_state.on_state_update()

        pass

    def switch_state(self, new_state: StateBase):
        """Properly switches the state machine's current state"""

        # Exit the current state
        if self.current_state is not None:
            self.current_state.on_state_exit()

        # Switch the current state to the new state
        self.current_state = new_state

        # Enter the new state
        if self.current_state is not None:
            self.current_state.on_state_enter()

        pass

    pass
